src/jit/hwintrinsiccodegenxarch.cpp

```diff
@@ -103,6 +111,243 @@ void CodeGen::genSSEIntrinsic(GenTreeHWIntrinsic* node)
+            // additional node in the chain.
+            assert(baseType == TYP_FLOAT);
+            assert(op1->OperIsList());
+            assert(op1->AsArgList()->Rest() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest()->Rest() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest()->Rest()->Rest() == nullptr);
+            assert(op2 == nullptr);
+
+            argList = op1->AsArgList();
+            op1     = argList->Current();
+            op1Reg  = op1->gtRegNum;
+            genConsumeRegs(op1);
+
+            argList = argList->Rest();
+            op2     = argList->Current();
+            op2Reg  = op2->gtRegNum;
+            genConsumeRegs(op2);
+
+            argList = argList->Rest();
+            op3     = argList->Current();
+            op3Reg  = op3->gtRegNum;
+            genConsumeRegs(op3);
+
+            argList = argList->Rest();
+            op4     = argList->Current();
+            op4Reg  = op4->gtRegNum;
+            genConsumeRegs(op4);
+
+            regNumber tmpReg = node->GetSingleTempReg();
+
+            emit->emitIns_SIMD_R_R_R(INS_unpcklps, tmpReg, op2Reg, op1Reg, TYP_SIMD16);
+            emit->emitIns_SIMD_R_R_R(INS_unpcklps, targetReg, op4Reg, op3Reg, TYP_SIMD16);
+            emit->emitIns_SIMD_R_R_R_I(INS_shufps, targetReg, targetReg, tmpReg, 68, TYP_SIMD16);
+            break;
+        }
+
+        case NI_SSE_SetAll:
+            assert(baseType == TYP_FLOAT);
+            assert(op2 == nullptr);
+            emit->emitIns_SIMD_R_R_R_I(INS_shufps, targetReg, op1Reg, op1Reg, 0, TYP_SIMD16);
+            break;
+
+        case NI_SSE_SetZero:
+            assert(baseType == TYP_FLOAT);
+            assert(op1 == nullptr);
+            assert(op2 == nullptr);
+            emit->emitIns_SIMD_R_R_R(INS_xorps, targetReg, targetReg, targetReg, TYP_SIMD16);
+            break;
+
+        case NI_SSE_Shuffle:
+        {
+            GenTreeArgList* argList;
+            ssize_t         ival;
+
+            // Shuffle takes 3 operands, so op1 should be an arg list with two
+            // additional node in the chain.
+            assert(baseType == TYP_FLOAT);
+            assert(op1->OperIsList());
+            assert(op1->AsArgList()->Rest() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest() != nullptr);
+            assert(op1->AsArgList()->Rest()->Rest()->Rest() == nullptr);
+            assert(op2 == nullptr);
+
+            argList = op1->AsArgList();
+            op1     = argList->Current();
+            op1Reg  = op1->gtRegNum;
+            genConsumeRegs(op1);
+
+            argList = argList->Rest();
+            op2     = argList->Current();
+            op2Reg  = op2->gtRegNum;
+            genConsumeRegs(op2);
+
+            argList = argList->Rest();
+            op3     = argList->Current();
+            ival    = op3->AsIntConCommon()->IconValue();
```

Comment:
This fails for the `Shuffle_r` test, but passes for the `Shuffle_ro` test. Seems to be because `op3` is a `GT_CAST` when optimizations are not enabled.

Reply:
👍

Reply:
Looks like this is because we aren't forcing expansion of the intrinsics, which causes other downstream processing on the args to occur.

Reply:
https://github.com/dotnet/coreclr/pull/15639